app/adapters/osu_beatmap_file.py

```
@@ -0,0 +1,508 @@
+    """The ``Mode`` field as a gamemode int.
+
+    Absent means osu!standard, which is how pre-v4 files express it. Anything
+    outside 0..3 raises rather than clamping, because a mode we do not know is
+    one we cannot rate.
+    """
+    if raw_mode is None or not raw_mode.strip():
+        return 0  # absent means osu!standard
+    try:
+        mode = int(float(raw_mode))
+    except ValueError:
+        raise OsuFileParseError(f"invalid Mode value {raw_mode!r}") from None
```

Ruff (0.16.2)

[warning] Avoid specifying long messages outside the exception class

(TRY003)